addons/product/tests/test_product_attribute_value_config.py

```diff
@@ -383,9 +385,10 @@ def test_get_first_possible_combination(self):
 
         # No possible combination (test helper and iterator)
         self._add_exclude(computer_ram_32, computer_hdd_4)
-        self.assertEqual(self.computer._get_first_possible_combination(), self.env['product.template.attribute.value'])
+        with self.assertRaises(MissingError, msg='The combinations should have been deleted'):
```

Comment:
It's not ok to change all these assertions. The code is not supposed to suddenly start raising errors.

Reply:
What would you recommend? I could check the number of combinations left such as in 
https://github.com/odoo/odoo/blob/86ce3dad19e5d1fbba4b1568cdd20ab0fce40c04/addons/product/tests/test_variants.py#L1314

which was my first thought. But then, you don't specifically evaluate in line 226 the combination `computer_ssd_512 + computer_ram_16 + computer_hdd_1` and `computer_ssd_256 + computer_ram_16 + computer_hdd_1`.

I don't know which is best.

Reply:
I don't understand what you are trying to do with these changes. This is testing that when there is no combination `_get_first_possible_combination` is returning the empty record set. It should not raise, if it does there is an error in your code.

Reply:
`self.computer._get_variant_for_combination(computer_ssd_256 + computer_ram_16 + computer_hdd_1)` will return en empty `product.product` but `_is_variant_possible` will raise an error because it won't pass `ensure_one`
https://github.com/odoo/odoo/blob/28f87fa620327ea7d06f641c868e7a50178e5526/addons/product/models/product.py#L695

Which is the behaviour that we want. 

Again, I could check whether `_get_variant_combination` return something instead -> that is something like:
```diff
--- a/addons/product/tests/test_product_attribute_value_config.py
+++ b/addons/product/tests/test_product_attribute_value_config.py
@@ -223,8 +223,7 @@ class TestProductAttributeValueConfig(TestProductAttributeValueSetup):

-        with self.assertRaises(ValueError, msg="Due to the exclusion it should not be possible"):
-            self.computer._get_variant_for_combination(computer_ssd_256 + computer_ram_16 + computer_hdd_1)._is_variant_possible()
+        self.assertIsEqual(len(self.computer._get_variant_for_combination(computer_ssd_256 + computer_ram_16 + computer_hdd_1)), 0)
```

What do you think?

Reply:
`_is_variant_possible` should not be called on a non existing variant. I'm not sure where it is done, but that needs to be changed.

Reply:
Hi @seb-odoo 

The issue happening since the last variant is removed, the product.template itself is removed:

https://github.com/odoo/odoo/blob/44f7754af4fc82228f4f12d3f0b59a2cfe7ecab5/addons/product/models/product.py#L350-L359

so what is missing is the product.template itself (`self.computer`) but this could just be tested with:

```py
self.assertFalse(self.computer.exists())
```

Reply:
This seems very unforgiving for the user.

That means if anybody adds an unfortunate exclusion rule that would make no variant possible, the template would suddenly be lost?

That might be a hard stop for anything we try to do in this PR then.

Reply:
🤔 Maybe a non blocking constraint can be added in the exclusion rule model, that checks every time a new exclusion is going to be added. The check would be as follows: if the new exclusion added would make all variants of a template be deleted (and thus make a template be deleted), then show pop up warning message, notifying the user that if they proceed and create the exclusion, then the xxx templates will be deleted.

Reply:
I'm still investigating but the last test in `test_variants.py` shows that if we exclude all variants the same way as a user would do in the interface it does not delete the `product.template` (I also tried it manually).

![image](https://user-images.githubusercontent.com/94046269/152172703-410482d7-5e62-4668-86b9-1ca01e7c3d27.png)

vs

![image](https://user-images.githubusercontent.com/94046269/152172841-6ed88095-13b0-4fd1-8994-4d8c263d903f.png)

Reply:
It depends on `has_dynamic_attributes`.

Reply:
Which is not the case for the attributes (e.g.: self.ram_16) of self.computer, isn't it?

Reply:
If the tests in `test_variants.py` passes, it is because we unlink multiple variants at one. The deletion of the template occurs only and only when there is **one** variant left and that a last exclusion rule deletes this last variant.